starfish/pipeline/features/codebook.py

```diff
@@ -0,0 +1,542 @@
+          * gene_name  (gene_name) object 'ACTB_human' 'ACTB_mouse'
+          * c          (c) int64 0 1 2 3
+          * h          (h) int64 0 1
+
+        Returns
+        -------
+        Codebook :
+            Codebook with shape (genes, channels, hybridization_rounds)
+
+        """
+        with open(json_codebook, 'r') as f:
+            code_array = json.load(f)
+        return cls.from_code_array(code_array, n_hyb, n_ch)
+
+    def save(self, filename: str) -> None:
+        """save a codebook to json
+
+        Notes
+        -----
+        This enforces the following typing of codebooks:
+        ch, hyb: int
+        value: float
+        gene: str
+
+        Parameters
+        ----------
+        filename : str
+            filename
+
+        """
+        code_array = []
+        for gene in self[self.Constants.GENE.value]:
+            codeword = []
+            for ch in self[Indices.CH.value]:
+                for hyb in self[Indices.HYB.value]:
+                    if self.loc[gene, ch, hyb]:
+                        codeword.append(
+                            {
+                                Indices.CH.value: int(ch),
+                                Indices.HYB.value: int(hyb),
+                                self.Constants.VALUE.value: float(self.loc[gene, ch, hyb])
+                            })
+            code_array.append({
+                self.Constants.CODEWORD.value: codeword,
+                self.Constants.GENE.value: str(gene.values)
+            })
+
+        with open(filename, 'w') as f:
+            json.dump(code_array, f)
+
+    def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
+        """Assign the closest gene by euclidean distance to each feature in an intensity table
+
+        Parameters
+        ----------
+        intensities : IntensityTable
+            features to be decoded
+
+        Returns
+        -------
+        IntensityTable :
+            intensity table containing additional data variables for gene assignments and feature
+            qualities
+
+        """
+
+        def _min_euclidean_distance(observation: xr.DataArray, codes: Codebook) -> np.ndarray:
+            """find the code with the closest euclidean distance to observation
+
+            Parameters
+            ----------
+            observation : xr.DataArray
+                2-dimensional DataArray of shape (n_ch, n_hyb)
+            codes :
+                Codebook containing codes to compare to observation
+
+            Returns
+            -------
+            np.ndarray :
+                1-d vector containing the distance of each code to observation
+
+            """
+            squared_diff = (codes - observation) ** 2
+            code_distances = np.sqrt(squared_diff.sum((Indices.CH, Indices.HYB)))
+            # order of codes changes here (automated sorting on the reshaping?)
+            return code_distances
+
+        # normalize both the intensities and the codebook
```

Comment:
i think this normalization step should be exposed as optional, but default to true for both codebook and intensity tables

Reply:
Good call. I've already fixed this in the MERFISH branch that I'm working on, which utilizes this function. I've Made #332 to remind myself.